File: apps/api/src/lib/embeddings.ts
import {
  LlmConfigurationError,
  LlmProviderError,
  openAiBaseUrl,
  resolveOpenAiKey,
} from "./openai";
import { deadlineSignal } from "./request-timeout";
import { EMBEDDING_DIMENSIONS, EmbeddingValidationError, resolveEmbeddingConfig, validateEmbedding, type EmbeddingConfig } from "./embedding-contract";

/**
 * 検索クエリの埋め込みベクトルを生成する。
 * OpenAI / Ollama both use the fixed 1536-dimensional storage contract.
 */

const DEFAULT_OLLAMA_BASE_URL = "http://127.0.0.1:11434";
const EMBEDDING_TIMEOUT_MS = 30_000;

export type EmbeddingEnv = {
  EMBEDDING_PROVIDER?: string;
  EMBEDDING_MODEL?: string;
  OPENAI_API_KEY?: string;
  OPENAI_BASE_URL?: string;
  OLLAMA_BASE_URL?: string;
};

async function embedWithOpenAi(env: EmbeddingEnv, text: string, config: EmbeddingConfig, signal?: AbortSignal): Promise<number[]> {
  const apiKey = resolveOpenAiKey(env, "OpenAI embeddings");
  const body: Record<string, unknown> = {
    model: config.model,
    input: text,
    encoding_format: "float",
    dimensions: EMBEDDING_DIMENSIONS,
  };

  let res: Response;
  try {
    res = await fetch(`${openAiBaseUrl(env)}/embeddings`, {
      method: "POST",
      headers: {
        "content-type": "application/json",
        authorization: `Bearer ${apiKey}`,
      },
      body: JSON.stringify(body),
      signal: deadlineSignal(EMBEDDING_TIMEOUT_MS, signal),
    });
  } catch {
    signal?.throwIfAborted();
    throw new LlmProviderError("OpenAI embeddings request failed. Check the server connection.");
  }
  if (!res.ok) {
    await res.body?.cancel();
    if (res.status === 401) throw new LlmConfigurationError("Invalid OpenAI API key. Check the server configuration.");
    throw new LlmProviderError(`OpenAI embeddings failed (HTTP ${res.status}).`);
  }

  const json: unknown = await embeddingJson(res, config, signal);
  if (!isRecord(json) || !Array.isArray(json.data) || json.data.length !== 1 ||
      !isRecord(json.data[0]) || json.data[0].index !== 0) {
    throw new EmbeddingValidationError("EMBEDDING_OUTPUT_INVALID", config);
  }
  return validateEmbedding(json.data[0].embedding, "EMBEDDING_OUTPUT_INVALID", config);
}

async function embedWithOllama(env: EmbeddingEnv, text: string, config: EmbeddingConfig, signal?: AbortSignal): Promise<number[]> {
  const base = (env.OLLAMA_BASE_URL || DEFAULT_OLLAMA_BASE_URL).replace(/\/$/, "");
  let res: Response;
  try {
    res = await fetch(`${base}/api/embed`, {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify({ model: config.model, input: text, dimensions: EMBEDDING_DIMENSIONS }),
      signal: deadlineSignal(EMBEDDING_TIMEOUT_MS, signal),
    });
  } catch {
    signal?.throwIfAborted();
    throw new LlmProviderError("Ollama embeddings request failed. Check the server connection.");
  }
  if (!res.ok) {
    await res.body?.cancel();
    throw new LlmProviderError(`Ollama embeddings failed (HTTP ${res.status}).`);
  }
  const json: unknown = await embeddingJson(res, config, signal);
  if (!isRecord(json) || !Array.isArray(json.embeddings) || json.embeddings.length !== 1) {
    throw new EmbeddingValidationError("EMBEDDING_OUTPUT_INVALID", config);
  }
  return validateEmbedding(json.embeddings[0], "EMBEDDING_OUTPUT_INVALID", config);
}

export async function embedQuery(env: EmbeddingEnv, text: string, signal?: AbortSignal): Promise<number[]> {
  signal?.throwIfAborted();
  const config = resolveEmbeddingConfig(env);
  const result = await (config.provider === "ollama"
    ? embedWithOllama(env, text, config, signal)
    : embedWithOpenAi(env, text, config, signal));
  signal?.throwIfAborted();
  return result;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

async function embeddingJson(res: Response, config: EmbeddingConfig, signal?: AbortSignal): Promise<unknown> {
  try { return await res.json(); }
  catch {
    signal?.throwIfAborted();
    throw new EmbeddingValidationError("EMBEDDING_OUTPUT_INVALID", config);
  }
}

File: apps/worker/worker_python/pipeline/embeddings.py
"""VideoQ embedding providers shared with the API (OpenAI / Ollama)."""

from __future__ import annotations

import json
import logging
import urllib.error
import urllib.request

from worker_python.env import env_str
from .embedding_contract import (
    EMBEDDING_DIMENSIONS,
    EmbeddingConfig,
    invalid_output,
    resolve_embedding_config,
    validate_embedding,
)

logger = logging.getLogger(__name__)


def embed_texts(texts: list[str]) -> list[list[float]]:
    if not texts:
        return []
    config = resolve_embedding_config()
    if config.provider == "ollama":
        return _embed_ollama(texts, config)
    return _embed_openai_batch(texts, config)


def _embed_openai_batch(texts: list[str], config: EmbeddingConfig) -> list[list[float]]:
    api_key = env_str("OPENAI_API_KEY")
    if not api_key:
        raise RuntimeError("OPENAI_API_KEY is required for OpenAI embeddings")
    base = env_str("OPENAI_BASE_URL", "https://api.openai.com/v1").rstrip("/")
    body: dict = {"model": config.model, "input": texts, "encoding_format": "float", "dimensions": EMBEDDING_DIMENSIONS}

    req = urllib.request.Request(
        f"{base}/embeddings",
        data=json.dumps(body).encode("utf-8"),
        headers={
            "content-type": "application/json",
            "authorization": f"Bearer {api_key}",
        },
        method="POST",
    )
    try:
        with urllib.request.urlopen(req, timeout=120) as resp:
            payload = _read_payload(resp, config)
    except urllib.error.HTTPError as exc:
        raise RuntimeError(f"OpenAI embeddings failed (HTTP {exc.code}).") from None
    except urllib.error.URLError:
        raise RuntimeError("OpenAI embeddings request failed. Check the server connection.") from None

    data = payload.get("data") if isinstance(payload, dict) else None
    if not isinstance(data, list) or len(data) != len(texts) or not all(
        isinstance(item, dict) and type(item.get("index")) is int for item in data
    ) or sorted(item["index"] for item in data) != list(range(len(texts))):
        raise invalid_output(config)
    return [validate_embedding(item.get("embedding"), config) for item in sorted(data, key=lambda item: item["index"])]


def _embed_ollama(texts: list[str], config: EmbeddingConfig) -> list[list[float]]:
    base = env_str("OLLAMA_BASE_URL", "http://127.0.0.1:11434").rstrip("/")
    req = urllib.request.Request(
        f"{base}/api/embed",
        data=json.dumps({"model": config.model, "input": texts, "dimensions": EMBEDDING_DIMENSIONS}).encode("utf-8"),
        headers={"content-type": "application/json"},
        method="POST",
    )
    try:
        with urllib.request.urlopen(req, timeout=120) as resp:
            payload = _read_payload(resp, config)
    except urllib.error.HTTPError as exc:
        raise RuntimeError(f"Ollama embeddings failed (HTTP {exc.code}).") from None
    except urllib.error.URLError:
        raise RuntimeError("Ollama embeddings request failed. Check the server connection.") from None
    vectors = payload.get("embeddings") if isinstance(payload, dict) else None
    if not isinstance(vectors, list) or len(vectors) != len(texts):
        raise invalid_output(config)
    return [validate_embedding(vector, config) for vector in vectors]


def _read_payload(response, config: EmbeddingConfig):
    try:
        return json.loads(response.read().decode("utf-8"))
    except (ValueError, UnicodeError):
        raise invalid_output(config) from None


def to_vector_literal(embedding: list[float]) -> str:
    return "[" + ",".join(str(float(x)) for x in embedding) + "]"
